### ISSUE TITLE
What signature should the absurdibles have?

### ISSUE THREAD

**Issue by treeowl** [leftaroundabout/trivial-constraint#5]
In addition to versions taking multiple proxies, it would be possible to add ones taking just one proxy, and none.

```haskell
absurdible2' :: proxy '(t, s) -> a
-- with AllowAmbiguousTypes
absurdible2'' :: a
```

Both of these seem more ergonomical than `absurdible2`.

But thinking again, it seems there's an even easier way.

```haskell
import Data.Constraint (Bottom (..))
class (TypeError ..., Bottom) => Impossible[n] ...
```

It doesn't really matter which instance, or even which *class* is used; whenever one of them is in play, `Data.Constraint.no` will do the trick.

**Comment by leftaroundabout** [comment]
Other suggestions you made:

- #3: All the proxy types are forced to be the same, and therefore all the class arguments are forced to have the same kind. They should take `proxyt t`, `proxys s`, etc.
- #4: The result of absurdible, etc., can be made levity polymorphic, so it probably should be.

**Comment by leftaroundabout** [comment]
@mstksg what do you think about this? The `absurdible` methods were your idea. I personally avoid using anything with proxies at all, IMO what's only really relevant for the type level should be kept in the type language.

The `proxy '(t, s)` suggestion sounds best to me: if we use proxies at all, at least only one. Plus, this could be made symmetrical: the `Unconstrained𝑁` class could have
```
  trivial𝑁 :: Proxy '(t, ...)
```

**Comment by leftaroundabout** [comment]
As for levity polymorphism: yeah, why not; can you make a PR for that? I still haven't learned how to do this.

**Comment by leftaroundabout** [comment]
And, could you sketch out a bit more what you mean by the `Bottom` suggestion?

**Comment by treeowl** [comment]
Yeah, sorry. I think the `Bottom` approach is probably the simplest/best way. We don't really *need* a bunch of `absurdible` methods, I don't think. We just need to ensure that whenever we know `Impossible[n] ...`, we can get whatever we want. One way to do that is to stick it in a superclass constraint! `Data.Constraint.Bottom` is a sort of `Impossible0`, except that it has no `TypeError` constraint (it instead has an `Any` constraint) and it's `absurdible0` is called `no`. When you're in a situation where you're given some impossibility and you need to produce something, you can just say `no` (or a levity polymorphic version thereof) and be done with it.

**Comment by leftaroundabout** [comment]
Ah, I see. No proxies at all then!

Well, I'm quite ok with that. But then, I myself don't need the `absurdible` methods _at all_ even now, I just see them as “we can add this and it directly shows what's going on and that the type `t` _can't exist_” – i.e. more as documentation / proof than something to actually use. But perhaps Justin sees that differently, so I'll wait what he has to say.

**Comment by treeowl** [comment]
As for the levity polymorphism, maybe that's better left upstream, or we can write

```haskell
nope :: forall (a :: TYPE rep). Bottom => a
nope = case no of
```

**Comment by treeowl** [comment]
As for using `no`: in a context where you would reach for `undefined`, relying on your knowledge that you're in an impossible context, you can just drop in `no` (or `nope`) instead and have the compiler prove safety for you.

**Comment by treeowl** [comment]
For example:

```haskell
class Foo c a where
  foo :: c b => proxy c -> a -> b

instance Foo Impossible a where
  foo = nope
```

No more code is needed to do it "right" than to throw in an `undefined`.

**Comment by mstksg** [comment]
I like using a master `Bottom`/`Impossible0` to tie everything together.  The idea of making them all superclasses is a nice one.

The main concern I had when designing this was the error message story.  I played around with it a bit and saw that adding something like `Bottom` as a constraint only adds an extra error message on top of the current readable one:

```
    * All instances of Impossible are disallowed
    * No instance for Bottom arising from the superclass constraint of an instance declaration
```

so the user would likely know exactly what is going on from the first message.

It does feel a bit overkill to require the entire constraints library just for a single typeclass, so defining a local `Impossible0` or `Bottom` might be preferred?

That being said, usability-wise, this fits with all of my current use cases of `Impossible` and `absurdible`, so I wouldn't have any problems usability-wise with these changes; the fact that they are now more flexible (no more overly constrained proxies and levity-polymorphism on the output) is a benefit without any drawback I can think of.

**Comment by mstksg** [comment]
Would you like to submit the PR @treeowl / @leftaroundabout ?  Otherwise I can get around to it some time later this week as well.

**Comment by leftaroundabout** [comment]
Would it have any advantage to use the same `Bottom` as `constraints` does? I wouldn't be _that_ worried about depending on `constraints`, but you're right: it's a quite heavy dependency for this itself ultra-slim package. So if there's no good reason, we could just port `Bottom` here, make it levity-polymorphic while we're at it, and use it as the superclass for all the `Impossible`s. The absurdibles could then be removed right again, if you're ok with that.

(Then later, it might actually make sense to file a PR over at `constraints`, so that _they_ should import `trivial-constraint` and its `Bottom` class.)

**Comment by mstksg** [comment]
I don't really see too much advantage to using `Bottom` from *constraints*, since I don't think there is that much of an ecosystem in haskell built around `Bottom`.  That being said, *constraints* itself doesn't depend on too much outside of what is already included with GHC, so either situation doesn't seem too bad.

**Comment by treeowl** [comment]
I don't care much either way.

**Comment by leftaroundabout** [comment]
https://github.com/leftaroundabout/trivial-constraint/tree/feature/unified-absurdible-method

Comments?

**Comment by leftaroundabout** [comment]
Do we actually need `Any` there? The levity-polymorphic `nope` actually makes an instance impossible to define already, doesn't it?

**Comment by treeowl** [comment]
We don't need `Any`, but for a different reason: there's a `TypeError` in the superclass constraint. The poly-kinded `nope` doesn't help directly, because `undefined` has that type.

**Comment by treeowl** [comment]
While you're at it, I think you might as well add an `Impossible0` class for completeness.

**Comment by leftaroundabout** [comment]
073cfa7

**Comment by leftaroundabout** [comment]
@mstksg are you also fine with this?

**Comment by mstksg** [comment]
Looks all good to me :)

**Comment by treeowl** [comment]
I'd also change the documentation of `Impossible` to indicate that a disabled method can be defined as `nope` rather than saying it can be left undefined. A full example might be nice.

**Comment by leftaroundabout** [comment]
Hm. Do you have a good example?

**Comment by treeowl** [comment]
Something like what I mentioned earlier should do:

```haskell
class C c a | a -> c where
  m :: c b => a -> b
  n :: Dict (c a) -> Int#
  ...

instance C Impossible Int where
  m = nope
  n Dict = nope -- Demonstrates levity polymorphism
  ...
```

**Comment by treeowl** [comment]
In case you're following by email, I've edited my last comment a couple times.

**Comment by leftaroundabout** [comment]
That Fundep type class indeed a nice example. If you want make a PR so you're the committer, else I'll include it tomorrow.

**Comment by leftaroundabout** [comment]
One more thing I wonder if somebody has an idea about: would it make any sense to add something dual to `nope`? My first though was that this would be simply
```
yup :: ()
yup = ()
```
...which wouldn't be very poignant. But maybe there is actually more to this: `nope` witnesses that the context is so constrained that it can never be fulfilled. Likewise, `yup` would supposedly witness that _the context does not constrain anything_. So perhaps it would actually be more like
```
yup :: (∀ c . c => a) -> a
```
...or, hm, I'm not sure what I really mean here.

Does that though make any sense at all?

**Comment by treeowl** [comment]
I don't think you can really go anywhere with `Unconstrained`. Maybe

```haskell
yup :: c :- Unconstrained0
```

gets the point across?
